src/autopas/selectors/AutoTuner.h

```diff
@@ -452,120 +510,197 @@ bool AutoTuner<Particle>::iteratePairwise(PairwiseFunctor *f, bool doListRebuild
-        auto &particleBufferInner = particleBuffers[bufferIdInner];
-        for (size_t j = bufferIdOuter == bufferIdInner ? i + 1 : 0; j < particleBufferInner.size(); ++j) {
-          auto &&p2 = particleBufferInner[j];
-          if (newton3) {
-            f->AoSFunctor(p1, p2, true);
-          } else {
-            f->AoSFunctor(p1, p2, false);
-            f->AoSFunctor(p2, p1, false);
-          }
+
+  // All (halo-)buffer interactions shall happen vectorized, hence, load all buffer data into SoAs
+  for (auto &buffer : particleBuffers) {
+    f->SoALoader(buffer, buffer._particleSoABuffer, 0);
+  }
+  for (auto &buffer : haloParticleBuffers) {
+    f->SoALoader(buffer, buffer._particleSoABuffer, 0);
+  }
+
+#ifdef AUTOPAS_OPENMP
+  // Tasks would be better than locks but sanitizers seem to have problems
+#pragma omp parallel
+#endif
+  {
+#ifdef AUTOPAS_OPENMP
+    // No need to synchronize after this loop since we have locks
+#pragma omp for nowait
+#endif
+    for (size_t i = 0; i < particleBuffers.size(); ++i) {
+      auto *particleBufferSoAA = &particleBuffers[i]._particleSoABuffer;
+      const std::lock_guard<std::mutex> lockA(*_bufferLocks[i]);
+      f->SoAFunctorSingle(*particleBufferSoAA, newton3);
+    }
+    if constexpr (newton3) {
+#ifdef AUTOPAS_OPENMP
+      // collapse loops for better scheduling
+#pragma omp for collapse(2)
+#endif
+      for (size_t i = 0; i < particleBuffers.size(); ++i) {
+        for (size_t j = i + 1; j < particleBuffers.size(); ++j) {
+          auto *particleBufferSoAA = &particleBuffers[i]._particleSoABuffer;
+          auto *particleBufferSoAB = &particleBuffers[j]._particleSoABuffer;
+          const std::lock_guard<std::mutex> lockA(*_bufferLocks[i]);
+          const std::lock_guard<std::mutex> lockB(*_bufferLocks[j]);
+          f->SoAFunctorPair(*particleBufferSoAA, *particleBufferSoAB, true);
         }
       }
-    }
-  }
-  // 4. particleBuffer with haloParticleBuffer
-  for (size_t bufferIdOuter = 0; bufferIdOuter < particleBuffers.size(); ++bufferIdOuter) {
-    auto &particleBufferOuter = particleBuffers[bufferIdOuter];
-    for (size_t i = 0; i < particleBufferOuter.size(); ++i) {
-      auto &&p1 = particleBufferOuter[i];
-      for (size_t bufferIdInner = 0; bufferIdInner < haloParticleBuffers.size(); ++bufferIdInner) {
-        auto &haloBufferInner = haloParticleBuffers[bufferIdInner];
-        for (size_t j = 0; j < haloBufferInner.size(); ++j) {
-          auto &&p2 = haloBufferInner[j];
-          if (newton3) {
-            f->AoSFunctor(p1, p2, true);
+    } else {
+#ifdef AUTOPAS_OPENMP
+      // collapse loops for better scheduling
+#pragma omp for collapse(2)
+#endif
+      for (size_t i = 0; i < particleBuffers.size(); ++i) {
+        for (size_t jj = 0; jj < particleBuffers.size(); ++jj) {
+          auto *particleBufferSoAA = &particleBuffers[i]._particleSoABuffer;
+          const auto j = (i + jj) % particleBuffers.size();
```

Comment:
What is the reason for this? Creating diffusion for better scheduling?

Reply:
Yep. If we had a regular `i : 0 -> N` and `j : 0 -> N` loop all tasks that need the same `j` are created at the same time, increasing the probability of running into a lock.